lib/ex_ark/serdes/output_stream.ex

```diff
@@ -0,0 +1,103 @@
+defmodule ExArk.Serdes.OutputStream do
+  @moduledoc """
+  Output stream for serialization.
+  """
+
+  use TypedStruct
+
+  alias ExArk.Ir.Field
+  alias ExArk.Registry
+  alias ExArk.Types
+  alias ExArk.Types.Primitives
+
+  require Logger
+
+  @type failure :: {:error, any()}
+
+  typedstruct enforce: true do
+    field :bytes, binary(), default: <<>>
+    field :offset, non_neg_integer(), default: 0
+  end
+
+  @spec advance(t(), non_neg_integer()) :: t()
+  def advance(%__MODULE__{bytes: bytes, offset: offset} = stream, count) do
+    %{stream | bytes: bytes <> :binary.copy(<<0>>, count), offset: offset + count}
+  end
+
+  @spec append(t(), binary()) :: t()
+  def append(%__MODULE__{bytes: bytes, offset: offset} = stream, new_bytes) do
+    %{stream | bytes: bytes <> new_bytes, offset: offset + byte_size(new_bytes)}
+  end
+
+  @spec write(t(), Field.t(), any(), Registry.t()) :: {:ok, t()} | failure()
+  def write(%__MODULE__{} = stream, %Field{type: _type} = field, data, registry) do
+    write_field(stream, field, data, registry)
+  end
```

Comment:
It's a little weird that these functions are in the stream modules. I will probably eventually move these to the `(de)serialization.ex` files, and try to break those up a bit if they are too large. But that will be in some future PR.